### PR TITLE
Sow Referral

### PR DESCRIPTION
**Pull request by fr1jo**
# Summary

Introduce a referral system for Pod sowing that rewards referrers with a percentage of Pods when referred users sow Beans.

# Problem

The protocol currently has no mechanism to incentivize user acquisition and growth through referrals. There is no way for users who bring new participants to the protocol to benefit from their contributions to network effects.

# Solution

Implement a referral system that:
- Allows users to specify a referral address when sowing Beans
- Automatically mints additional Pods to the referrer based on a configurable percentage
- Tracks referral activity through events for analytics
- Integrates with the Tractor blueprint system for advanced sowing strategies

### Referral System Architecture

**Storage**
- Add `referralPercentage` to System storage (initialized to 1%)
- Configurable percentage determines referrer rewards

**Sowing Flow**
1. User sows Beans with optional referral address
2. Protocol burns Beans and mints Pods to user as normal
3. If referral address provided:
   - Calculate referral amount: `referralBeans = beans * referralPercentage / PRECISION`
   - Mint additional Pods to referrer at same temperature
   - Emit `ReferralSow` event for tracking

**Blueprint Integration**
- New `SowBlueprintv0Referral` contract extends Tractor functionality
- Supports complex sowing operations with referral tracking
- Includes conditions, withdrawal plans, and tip structures
- Enables automated, recurring sowing with referral benefits

### Contract Changes

**Modified Contracts:**
- `FieldFacet.sol` - Updated `sowWithMin` signature to accept referral address
- `LibDibbler.sol` - Core sowing logic updated to handle referral minting
- `System.sol` - Added `referralPercentage` storage variable
- `IBeanstalk.sol` - Updated interface to reflect new parameters

**New Contracts:**
- `SowBlueprintv0Referral.sol` - Tractor blueprint for sowing with referrals (510 lines)
- `InitPodReferral.sol` - Initialization contract for referral system deployment

**Tests:**
- Added comprehensive test coverage in `test/foundry/field/Field.t.sol`
- Tests cover basic referral sowing, blueprint integration, and edge cases

### Events

```solidity
event ReferralSow(
    address indexed account,
    address indexed referral,
    uint256 beans,
    uint256 pods
);
```

### Key Features

1. **No Additional Cost to Sower**: Referral Pods are minted from additional Bean supply, not from the sower's allocation
2. **Same Terms**: Referrer receives Pods at the same temperature as the original sow
3. **Flexible Integration**: Works with standard sowing and Tractor blueprints
4. **Configurable**: Referral percentage can be adjusted via governance
5. **Transparent**: All referral activity tracked via events

### Initial Parameters

- **Referral Percentage**: 1% (0.01e18)
- Can be updated through protocol governance

## Impact

- **User Growth**: Incentivizes organic user acquisition
- **Network Effects**: Rewards users who strengthen the protocol ecosystem
- **Backward Compatible**: Existing sowing functionality unchanged (referral is optional)
- **Gas Efficient**: Minimal overhead when not using referrals

### CODE REVIEW

File: contracts/beanstalk/init/InitPodReferral.sol [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * @notice Initializes the Pod referral system
+ * @dev Sets the referralPercentage in System storage
+ **/
+contract InitPodReferral {
+    /// @notice Emitted when the referral percentage is changed (with 1e18 precision)
+    event ReferralPercentageChanged(uint128 newReferrerPercentage);
+    event RefereePercentageChanged(uint128 newRefereePercentage);
+
+    uint128 internal constant INIT_REFERRER_PERCENTAGE = 0.01e18; // 10%
+    uint128 internal constant INIT_REFEREE_PERCENTAGE = 0.01e18; // 10%
```
Comment (PintoPirate) [inline contracts/beanstalk/init/InitPodReferral.sol]: intentional that 100% is 1e17?
